File: ProblemSolve/973.cpp
/******************************
 * K Closest Points to Origin *
 ******************************/

//참고 : https://leetcode.com/problems/k-closest-points-to-origin/discuss/217966/C%2B%2B-3-lines-nth_element-O(n)

class Solution {
public:
    //O(n)
    //nth_element를 이용한 해답
    //179ms, 49.2 MB
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {        
        //거리 순으로 k-1번째 index까지 정렬
        nth_element(points.begin(), points.begin() + k - 1, points.end(), 
                   [](vector<int>& a, vector<int>& b){
                       return (a[0] * a[0] + a[1] * a[1]) < (b[0] * b[0] + b[1] * b[1]);
                   });
        
        //필요 없는 부분 삭제
        points.resize(k);
        
        return points;
    }
    
    //O(n log n)
    //sort를 이용한 해답
    //277ms, 49.2 MB
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        
        //거리 순으로 정렬
        sort(points.begin(), points.end(), [](vector<int>& a, vector<int>& b){
                       return (a[0] * a[0] + a[1] * a[1]) < (b[0] * b[0] + b[1] * b[1]);
                   });
        
        
        //필요 없는 부분 삭제
        points.resize(k);
        
        return points;
    }
};

File: ProblemSolve/973.cs
/******************************
 * K Closest Points to Origin *
 ******************************/

public class Solution {
    public void Swap(ref int[] a, ref int[] b){
        int[] temp = a;
        a = b;
        b = temp;
    }

    public void QuickSort(ref int[][] points, int start, int end){
        if(start >= end){
            return;
        }

        ref int[] pivot = ref points[end];
        int left = start;
        int right;

        for(right = start; right<end; ++right){
            int dist1 = points[right][0] * points[right][0] + points[right][1] * points[right][1];
            int dist2 = pivot[0] * pivot[0] + pivot[1] * pivot[1];
            
            if(dist1 < dist2){
                Swap(ref points[right], ref points[left]);
                ++left;
            }
        }
        
        Swap(ref points[left], ref pivot);

	    QuickSort(ref points, left + 1, end);
	    QuickSort(ref points, start, left - 1);

	    return;
    }
    
    public int[][] KClosest(int[][] points, int k) {
        //Sorting
        QuickSort(ref points, 0, points.Length-1);
        
        Array.Resize(ref points, k);
        
        return points;
        
        //LinQ
        //return points.OrderBy(x => x[0] * x[0] + x[1] * x[1]).Take(k).ToArray();
    }
}